Applications/Spire/Source/OrderImbalanceIndicator/CachedOrderImbalanceIndicatorModel.cpp

```diff
@@ -0,0 +1,108 @@
+#include "Spire/OrderImbalanceIndicator/CachedOrderImbalanceIndicatorModel.hpp"
+#include "Spire/Spire/QtPromise.hpp"
+
+using namespace boost::icl;
+using namespace boost::posix_time;
+using namespace Nexus;
+using namespace Spire;
+
+CachedOrderImbalanceIndicatorModel::CachedOrderImbalanceIndicatorModel(
+  std::shared_ptr<OrderImbalanceIndicatorModel> m_source_model)
+  : m_source_model(std::move(m_source_model)) {}
+
+OrderImbalanceIndicatorModel::SubscriptionResult
+    CachedOrderImbalanceIndicatorModel::subscribe(
+    const boost::posix_time::ptime& start,
+    const boost::posix_time::ptime& end,
+    const OrderImbalanceSignal::slot_type& slot) {
+  if(contains(m_ranges, continuous_interval<ptime>::closed(start, end))) {
+    return get_subscription(start, end, slot);
+  }
+  return load_imbalances(start, end, slot);
```

Comment:
Look into only loading the portion that isn't cached.

Reply:
The loaded portions are calculated on line 47-48, where I take the difference of the requested range and the previously loaded ranges.